### DISCUSSION TITLE
useStore(state => state.slice): component not being re-rendered

### DISCUSSION THREAD

**Discussion by stephanedupont** [pmndrs/zustand#836] [Q&A] [answered]
Hi,

I'm currently working on a web app that has 3 apps inside. Let's call them `A`, `B` and `C`.

Each one of these three apps have a separate store, but with the same data model (and reducer).

```javascript
useAAppDataStore = create(redux(...))
useBAppDataStore = create(redux(...))
useCAppDataStore = create(redux(...))
```

There's a top-level menu that allows to select the current app, and and sub-menu (`AppMenu`) that has an app prop, and that renders differently depending on the app passed as a prop.

```javascript
function AppMenu({app}) {
  ...
  let bears;
  if (app === App.A) {
    bears = useAAppDataStore(state => state.bears);
  } else if (app === App.B) {
    bears = useBAppDataStore(state => state.bears);
  } else  if (app === App.C) {
    bears = useCAppDataStore(state => state.bears);
  }
  ...
}
```

I know that we are not supposed to use conditionals with hooks, but as for the same props of `AppMenu` there is always one and only one hook, I thought this would be ok.

But it appears that it doesn't work. I have some data sync occurring outside of react asynchronously. The sync starts when an app is selected in the top level menu. When the sync is done, it updates the app store (depending on the app) via its reducer.

For the first app that I select, it always works. The `AppMenu` is re-rendered if `state.bears` has been changed by the reducer.

But if I select another app it doesn't work. The store is updated by the reducer during the sync, but the `AppMenu` is not re-rendered.

If I change to:

```javascript
function AppMenu({app}) {
  ...
  const ABears = useAAppDataStore(state => state.bears);
  const BBears = useBAppDataStore(state => state.bears);
  const CBears = useCAppDataStore(state => state.bears);
  }
  ...
}
```

... then it works. But this is probably not a good solution as `AppMenu` only depends on the state of a single store, not all three of them.

I'm new to React (and therefore Zustand!), but how am I supposed to approach this? Any idea/tips/explanation about why this is happening and how to better implement this?

**Comment by dai-shi** [comment, marked as answer]
Conditional hooks don't work by design. In v4, we plan to provide a new API to cover this use case.
But, in this case, it should be generally easier to have just one store, instead of three.

**Comment by stephanedupont** [comment]
Thank you for your answer.

Wouldn't it be the same with a single store?

If I do something like this:

```javascript
function AppMenu({app}) {
  ...
  let bears;
  if (app === App.A) {
    bears = useAppDataStore(state => state.A.bears);
  } else if (app === App.B) {
    bears = useAppDataStore(state => state.B.bears);
  } else  if (app === App.C) {
    bears = useAppDataStore(state => state.C.bears);
  }
  ...
}
```

... it's sill a conditional hook isn't it?

**Reply by dai-shi** [reply]
Something like this:
```js
function AppMenu({app}) {
  ...
  const bears = useAppDataStore(state => {
    if (app === App.A) {
      return state.A.bears;
    } else if (app === App.B) {
      return state.B.bears;
    } else  if (app === App.C) {
      return state.C.bears;
    } else {
      throw new Error("...");
    }
  });
  ...
}
```

**Comment by stephanedupont** [comment]
OK, I think I understand.

Now, I would really like to keep the three stores separate.

You mentioned a v4 that would solve this without needing to merge the three stores into a single store.

Project has a few months of development left, so I'm not in a rush: any idea when v4 would be released approximately, and what the new API would look like?

**Reply by dai-shi** [reply]
https://www.npmjs.com/package/zustand/v/4.0.0-beta.2

When @devanshj completes #725, we will release `-rc`. After we collect feedbacks, **and** improve docs, **and** React 18 is released, v4 will be ready.

It should be something like this:
```js
import createStore from 'zustand/vanilla'
import { useStore } from 'zustand'

const aAppDataStore = createStore(...)
const bAppDataStore = createStore(...)
const cAppDataStore = createStore(...)

  // in component/hook
  let store
  if (app === App.A) {
    store = aAppDataStrore
  } else if (app === App.B) {
    store = bAppDataStore
  } else {
    store = cAppDataStore
  }
  const bears = useStore(store, state => state.bears)
```

Please try.

**Comment by stephanedupont** [comment]
That works great, thank you!

Should I use React 18 beta to go with Zustand 4? (Still at React 17.X, but it seems to work fine)

**Reply by dai-shi** [reply]
No, zustand v4 works just fine with React 17.

**Comment by stephanedupont** [comment]
Also, I have an additional fourth store for things that are global.

Will `createStore(...)` / `useStore(...)` be the standard way to use Zustand in v4? Or should I keep using `create()` for the global store if I don't need this specific use case?

**Reply by dai-shi** [reply]
Probably, the standard way is still `create()` for normal use cases.

**Comment by stephanedupont** [comment]
Thank you very much @dai-shi.
